### PR TITLE
Use new Cypher projections API when running against 2.4.0+

### PR DESCRIPTION
**Pull request by knutwalker**
Thank you for your contribution to the Graph Data Science Client project.

Before submitting this PR, please read [Contributing to the Neo4j Ecosystem](https://github.com/neo4j/graph-data-science-client/blob/main/CONTRIBUTING.md). 

Make sure:
- [x] You signed the [Neo4j CLA](https://neo4j.com/developer/cla/#sign-cla) (Contributor License Agreement) so that we are allowed to ship your code in our library
- [x] Your contribution is covered by tests

### CODE REVIEW

File: graphdatascience/query_runner/cypher_graph_constructor.py [hunk 1]
```diff
@@ -278,46 +306,52 @@ def collect_to_dict(row: Dict[str, Any]) -> Dict[str, Any]:
...
             nodes_config_fields: List[str] = []
-            if reduce(lambda x, y: x | y.has_labels(), node_cols, False):
+            if any(x.has_labels() for x in node_cols):
                 nodes_config_fields.append(
                     f"{CypherAggregationApi.SOURCE_NODE_LABEL}: {CypherAggregationApi.SOURCE_NODE_LABEL}"
                 )
+                if is_cypher_projection_v2:
+                    nodes_config_fields.append(
+                        f"{CypherAggregationApi.TARGET_NODE_LABEL}: NULL",
+                    )
```
Comment (knutwalker) [inline graphdatascience/query_runner/cypher_graph_constructor.py:329]: Any reason that we didn't set targetNodeLabels before? Should we? Or are we sure that all nodes with labels appear at least once in the sourceNode position?

File: graphdatascience/query_runner/cypher_graph_constructor.py [hunk 2]
```diff
@@ -278,46 +306,52 @@ def collect_to_dict(row: Dict[str, Any]) -> Dict[str, Any]:
...
 
             # as we first list all nodes at the top of the df, we don't need to lookup properties for the target node
-            if reduce(lambda x, y: x | y.has_properties(), node_cols, False):
+            if any(x.has_properties() for x in node_cols):
                 nodes_config_fields.append(
                     f"{CypherAggregationApi.SOURCE_NODE_PROPERTIES}: {CypherAggregationApi.SOURCE_NODE_PROPERTIES}"
                 )
+                if is_cypher_projection_v2:
```
Comment (knutwalker) [inline graphdatascience/query_runner/cypher_graph_constructor.py:336]: Same question about targetNodeProperties? Any reason that we didn't set targetNodeProperties before? Should we? Or are we sure that all nodes with properties appear at least once in the sourceNode position?

File: graphdatascience/query_runner/cypher_graph_constructor.py (hunk 1 above)
Comment (knutwalker) [reply graphdatascience/query_runner/cypher_graph_constructor.py:329]: cool, thanks!

File: graphdatascience/query_runner/cypher_graph_constructor.py [hunk 3]
```diff
@@ -120,11 +127,13 @@ def __init__(
             graph_name: str,
```
[the remark this answers is not in the document]
Comment (knutwalker) [reply graphdatascience/query_runner/cypher_graph_constructor.py:127]: There already is a CypherProjectionRunner… well, actually there is a CyperProjectionRunner ^^
The new projection is only used in 2.3+ when we added concurrency support.
I can follow the new naming and call then Legacy and v2.

File: graphdatascience/query_runner/cypher_graph_constructor.py [hunk 4]
```diff
@@ -137,8 +146,16 @@ def run(self, node_dfs: List[DataFrame], relationship_dfs: List[DataFrame]) -> N
...
+                CypherAggregationApi.REL_PROPERTIES_NEW
+                if is_cypher_projection_v2
+                else CypherAggregationApi.REL_PROPERTIES
+            )
+
+            aligned_node_dfs = self.adjust_node_dfs(node_dfs, graph_schema, properties_key)
+            aligned_rel_dfs = self.adjust_rel_dfs(relationship_dfs, graph_schema, properties_key)
 
             # concat instead of join as we want to first have all nodes and then the rels
             # this way we don't duplicate the node property data and its cheaper
```
[the remark this answers is not in the document]
Comment (knutwalker) [reply graphdatascience/query_runner/cypher_graph_constructor.py:161]: I think the wording is fine, I just never read that comment 😅

File: graphdatascience/query_runner/cypher_graph_constructor.py [hunk 5]
```diff
@@ -278,46 +306,52 @@ def collect_to_dict(row: Dict[str, Any]) -> Dict[str, Any]:
...
                     f"{CypherAggregationApi.SOURCE_NODE_LABEL}: {CypherAggregationApi.SOURCE_NODE_LABEL}"
                 )
+                if is_cypher_projection_v2:
+                    nodes_config_fields.append(
+                        f"{CypherAggregationApi.TARGET_NODE_LABEL}: NULL",
+                    )
 
             # as we first list all nodes at the top of the df, we don't need to lookup properties for the target node
-            if reduce(lambda x, y: x | y.has_properties(), node_cols, False):
+            if any(x.has_properties() for x in node_cols):
```
[the remark this answers is not in the document]
Comment (knutwalker) [reply graphdatascience/query_runner/cypher_graph_constructor.py:332]: There's also [all](https://docs.python.org/3.11/library/functions.html#all) for the other way :)

**Comment by knutwalker** [comment]
@FlorentinD thanks and 🏓

**Review by FlorentinD — APPROVED** [review]

**Pull request merged**